python/vsi/tools/mpl.py

```diff
@@ -0,0 +1,429 @@
+    else:
+      return 'Todo'
+
+  def cleanup(self, event):
+    self.event2 = event
+    if event.key == 'escape':
+      bubbles = self.get_bubbles(event.inaxes)
+      for bubble in bubbles:
+        event.inaxes.texts.remove(bubble)
+      bubbles.clear()
+
+  def _picker(self, event):
+    self.event = event # for debugging
+    if isinstance(event, mpl.backend_bases.PickEvent) and \
+       event.mouseevent.button == mpl.backend_bases.MouseButton.LEFT:
+      if isinstance(event.artist, mpl.lines.Line2D):
+        self.xdata = event.artist.get_xdata()
+        self.ydata = event.artist.get_ydata()
+        self.ind = event.ind
+        # self.x = np.take(self.xdata, self.ind)[0]
+        # self.y = np.take(self.ydata, self.ind)[0]
+        self.x = self.xdata[self.ind[0]]
+        self.y = self.ydata[self.ind[0]]
+
+        self.text_x = self.x
+        self.text_y = self.y
+      elif isinstance(event.artist, mpl.image.AxesImage):
+        self.image = event.artist.get_array()
+
+        extent = event.artist.get_extent()
+        x_scale = (extent[1]-extent[0]) / self.image.shape[1]
+        y_scale = (extent[3]-extent[2]) / self.image.shape[0]
+
+        if event.artist.origin == 'upper':
+          y_scale = -y_scale
+
+        if y_scale < 0:
+          y_offset = max(extent[3], extent[2])
+        else:
+          y_offset = min(extent[3], extent[2])
+
+        if x_scale < 0:
+          x_offset = max(extent[1], extent[0])
+        else:
+          x_offset = min(extent[1], extent[0])
+
+        self.x = math.floor((event.mouseevent.xdata-x_offset)/x_scale)
+        self.y = math.floor((event.mouseevent.ydata-y_offset)/y_scale)
+
+        self.text_x = (self.x + 0.5) * x_scale + x_offset
+        self.text_y = (self.y + 0.5) * y_scale + y_offset
+
+      elif isinstance(event.artist, mpl.collections.PathCollection): # scatter
+        self.xdata = event.artist.get_offsets().data[:,0]
+        self.ydata = event.artist.get_offsets().data[:,1]
+
+        self.ind = event.ind
+        self.x = self.xdata[self.ind[0]]
+        self.y = self.ydata[self.ind[0]]
+
+        self.text_x = self.x
+        self.text_y = self.y
+      elif isinstance(event.artist, mpl.collections.PatchCollection):
+        self.ind = event.ind
+        self.patches = event.artist.get_paths()
+        self.patch = self.patches[self.ind[0]]
+        self.text_x, self.text_y = self.patch.get_extents().corners().mean(axis=0)
+      elif isinstance(event.artist, mpl.text.Text):
+        self.x, self.y = event.artist.get_position()
+        self.text_x = self.x
+        self.text_y = self.y
+      elif isinstance(event.artist, mpl.patches.Rectangle):
+        self.bbox = event.artist.get_bbox()
+        # self.x, self.y = event.artist.get_bbox()
+        self.text_x = (self.bbox.x0 + self.bbox.x1)/2
+        self.text_y = (self.bbox.y0 + self.bbox.y1)/2
+      else:
+        plt.gca().set_title('oops')
```

Comment:
Should this raise an Unimplemented Exception? or just set the title to the empty string?

Reply:
Exceptions don't work here. They are unobservable. So I picks something that is observable.